components/nimbus/src/behavior.rs

```diff
@@ -190,8 +200,31 @@ impl SingleIntervalCounter {
         Self::new(config)
     }
 
-    pub fn increment(&mut self) -> Result<()> {
-        self.data.increment()
+    pub fn increment_then(&mut self, then: DateTime<Utc>, count: u64) -> Result<()> {
+        use std::cmp::Ordering;
+        let now = self.data.starting_instant;
+        let rotations = self.config.interval.num_rotations(then, now)?;
+        match rotations.cmp(&0) {
+            Ordering::Less => {
+                /* We can't increment in the future */
+                return Err(NimbusError::BehaviorError(BehaviorError::InvalidState(
+                    "Cannot increment events far into the future".to_string(),
+                )));
+            }
+            Ordering::Equal => {
+                if now < then {
+                    self.data.increment_at(0, count)?;
+                } else {
+                    self.data.increment_at(1, count)?;
+                }
+            }
+            Ordering::Greater => self.data.increment_at(1 + rotations as usize, count)?,
+        }
+        Ok(())
+    }
```

Comment:
This block, and the one above it with `increment_at` are the heart of this change.